controllers/limitador_controller.go

```diff
@@ -0,0 +1,152 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"github.com/3scale/limitador-operator/pkg/limitador"
+	v1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	limitadorv1alpha1 "github.com/3scale/limitador-operator/api/v1alpha1"
+)
+
+// LimitadorReconciler reconciles a Limitador object
+type LimitadorReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=limitador.3scale.net,resources=limitadors,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=limitador.3scale.net,resources=limitadors/status,verbs=get;update;patch
+
+func (r *LimitadorReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	reqLogger := r.Log.WithValues("limitador", req.NamespacedName)
+
+	// Delete Limitador deployment and service if needed
+	limitadorObj := limitadorv1alpha1.Limitador{}
+	if err := r.Get(context.TODO(), req.NamespacedName, &limitadorObj); err != nil {
+		if errors.IsNotFound(err) {
+			if err = r.ensureLimitadorDeploymentIsDeleted(req.NamespacedName); err != nil {
+				reqLogger.Error(err, "Failed to delete Limitador deployment.")
+				return ctrl.Result{}, err
+			}
+
+			if err = r.ensureLimitadorServiceIsDeleted(); err != nil {
+				reqLogger.Error(err, "Failed to delete Limitador service.")
+				return ctrl.Result{}, err
+			}
+
+			return ctrl.Result{}, nil
+		} else {
+			reqLogger.Error(err, "Failed to get Limitador object.")
+			return ctrl.Result{}, err
+		}
+	}
+
+	if err := r.ensureLimitadorServiceExists(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	desiredDeployment := limitador.LimitadorDeployment(&limitadorObj)
+
+	if err := r.reconcileDeployment(desiredDeployment); err != nil {
+		reqLogger.Error(err, "Failed to update Limitador deployment.")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *LimitadorReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&limitadorv1alpha1.Limitador{}).
+		Complete(r)
+}
+
+func (r *LimitadorReconciler) reconcileDeployment(desiredDeployment *v1.Deployment) error {
+	currentDeployment := v1.Deployment{}
+	key, _ := client.ObjectKeyFromObject(desiredDeployment)
+
+	err := r.Get(context.TODO(), key, &currentDeployment)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return r.Create(context.TODO(), desiredDeployment)
+		} else {
+			return err
+		}
+	}
+
+	if reflect.DeepEqual(currentDeployment.Spec, desiredDeployment.Spec) {
```

Comment:
Reconciling the whole `spec` section of some K8s objects might have surprising behaviours, be careful with this.
Some fields are automatically and explicitely set by K8s/OCP after being created, which lead to differences being detected.
For example, in a `Service` object, when it is created with type `ClusterIP` (the default behavior), its `.spec.clusterIP` field is automatically assigned and set an IP value. This would made the comparison detect a difference between your desired coded object and the existing one received from K8s.

Reply:
Good point. Changed :+1: